docs/spec.md

```diff
@@ -140,240 +43,355 @@ stablehlo.func @main(
+many programming languages, with the main peculiarity being StableHLO's
+domain-specific nature which results in some unusual outcomes (e.g. scalar types
+are not value types).
+
+```ebnf
+TensorType    ::= 'tensor' '<' TensorShape ElementType '>'
+TensorShape   ::= {DimensionSize 'x'}
+DimensionSize ::= digit {digit}
+```
 
-Within the grid, the programs can all be the same (in the "Single Program,
-Multiple Data" style), can all be different (in the "Multiple Program, Multiple
-Data" style) or something in between.
+**Tensor types** represent tensors, i.e. multidimensional arrays. They have a
+**shape** and an **element type**, where a shape represents non-negative
+**dimension sizes** in the ascending order of the corresponding **dimensions**
+(which are also called **axes**) numbered from `0` to `R-1`. The number of
+dimensions `R` is called **rank**. For example, `tensor<2x3xf32>` is a tensor
+type with shape `2x3` and element type `f32`. It has two dimensions (or,
+in other words, two axes) - 0th dimension and 1st dimension - whose sizes are
+2 and 3. Its rank is 2.
+
+```ebnf
+TokenType ::= 'token'
+```
 
-Within the grid, the processes are mostly independent from each other - they
-have separate operation statuses, separate input/intermediate/output values and
-most of the ops are executed separately between processes, with the exception of
-a small number of collective ops described below.
+**Token types** represent tokens, i.e. opaque values produced and consumed
+by some operations. Tokens are used for imposing execution order on operations
+as described in the "Execution" section.
 
-Given that execution of most of the ops is only using values from the same
-process, it is usually unambiguous to refer to these values by their names.
-However, when describing semantics of collective ops, that is insufficient, and
-we use the notation `name@process_id` to refer to the value `name` within a
-particular process. (From that perspective, unqualified `name` can be viewed as
-a shorthand for `name@(replica_id(), partition_id())`).
+```ebnf
+TupleType ::= 'tuple' '<' [ValueType {',' ValueType}] '>'
+```
 
-The execution order across processes is implementation-defined, except for the
-synchronization introduced by point-to-point communication and collective ops
-as described below.
+**Tuple types** represent tuples, i.e. heterogeneous lists. Tuples are a legacy
+feature which only exists for compatibility with HLO. In HLO, tuples are
+used to represent variadic inputs and outputs. In StableHLO, variadic inputs and
+outputs are supported natively, and the only use of tuples in StableHLO is to
+comprehensively represent HLO ABI where e.g. `T`, `tuple<T>` and
+`tuple<tuple<T>>` may be materially different depending on a particular
+implementation.
+
+```ebnf
+ElementType ::= BooleanType | IntegerType | FloatType | ComplexType
+BooleanType ::= 'i1'
+IntegerType ::= 'si4' | 'si8' | 'si16' | 'si32' | 'si64'
```

Comment:
We support `4/8/16/32/64-bit signless integer or 4/8/16/32/64-bit unsigned integer`.

This should probably read `'i4' | 'i8' | 'i16' | 'i32' | 'i64'` as we error on `si`.